### ISSUE TITLE
[ttnn/topk] NaN lanes are ordered by sign+magnitude of the raw bit pattern: a -NaN ranks below -inf and silently drops out of the top-k (fp32 as well as bf16)

### ISSUE THREAD

**Issue by heroxcreed** [tenstorrent/tt-metal#56626]
**Component / Area:** ttnn reduction - `ttnn.topk` (the value-lane ordering produced by the SFPU local-sort / merge network)
**Issue Type:** Bad Outputs

## Summary

`ttnn.topk` does not order NaN lanes the way its own documentation and registered golden say it should. On the released 0.78.0 wheel, value lanes are ordered by a **(sign, magnitude) rule applied to the raw bit pattern**, with no NaN case at all. A NaN therefore sorts by its mantissa, which is larger than an infinity's:

* `+NaN` (`0x7FC0` / `0x7FC00000`) ranks **above** `+inf`. For `largest=True` that coincidentally agrees with `torch.topk`.
* `-NaN` (`0xFFC0` / `0xFFC00000`) ranks **below** `-inf`, i.e. below every finite value and below the smallest representable number.

Consequence, `largest=True`, `k=16`, in a row `1..64` with one `-NaN` lane: the NaN is not merely re-encoded, it **drops out of the top-k entirely**, so the op returns a different *set* of elements than `torch.topk` - 16/16 values and 16/16 indices differ. Two NaNs that differ only in their sign bit end up at the two opposite extremes of the row.

For the identical input in bf16 the **permutation is the same** (same indices as fp32); only the returned values are additionally flushed to infinities. That value substitution is the known bf16 packer family (#51557 / #54381 / #56414) and is **not** the subject of this report - see "What this is not".

### The contract, quoted from the op itself

* `ttnn.topk.__doc__` (0.78.0 wheel): "Equivalent PyTorch code: `return torch.topk(input_tensor, k, dim=dim, largest=largest, sorted=sorted, *, output_tensor=None)`".
* The registered golden is literally `torch.topk`: `ttnn/operations/reduction.py`, `_create_golden_function_topk` -> `return torch.topk(input_tensor, k, dim=dim, largest=largest, sorted=sorted)`.
* In-repo precedent that "NaN is the largest element" is the intended contract for a selection reduction: #55346, fixed by PR #55350, which aligned the `argmax` comparators with torch NaN and signed-zero semantics.

## Environment

* `ttnn` python wheel **0.78.0**; bundled `tt-llk` topk header md5 `cfae670a7558c946e4d5bd77df7b4867` (unmodified, backup/restore/verify checked).
* **TT simulator (ttsim) only - no silicon attached**: `TT_METAL_SIMULATOR=$HOME/sim/libttsim_bh.so`, `TT_METAL_SLOW_DISPATCH_MODE=1`, `TT_METAL_DISABLE_SFPLOADMACRO=1`.
* `TILE_LAYOUT` on device, shape `[1,1,1,64]`, `dim=3`, `k=16/32/64`, `largest=True` and `False`.
* Every NaN/Inf is written as a raw bit pattern through `tensor.view(torch.int16|torch.int32)` and every result is read back as raw bits. This matters: a NaN passed through a python float is silently rewritten to `0xFFFF` (bf16) / `0xFFFF0000` (fp32) on this platform, which destroys any sign-based conclusion. The reproducer below also reads the input tensor back **from the device** and prints the sent patterns, so the input is confirmed before the op runs.

**Environment limitation:** everything here is the ttsim model of the compute core, not silicon. Also, the `ttnn.topk` stack has been rewritten since 0.78.0 (current main uses the fused-key local-sort, tt-llk header `4daf16fc`, 1676 lines) and I cannot build/run main in this environment, so "affects main too" is source-based inference, not a measurement (last section).

## Evidence (bit-exact, single run of the reproducer below)

Input row: lanes 0..59 = `1.0 .. 60.0`, lane 60 = `+NaN`, lane 61 = `-NaN`, lane 62 = `+inf`, lane 63 = `-inf`.

**A. The sent patterns do reach the device tensor** (`from_torch` -> `to_torch` -> bits):

```
[bf16] roundtrip lanes 60:64 = ['7fc0', 'ffc0', '7f80', 'ff80']
[fp32] roundtrip lanes 60:64 = ['7fc00000', 'ffc00000', '7f800000', 'ff800000']
```

**B. `largest=True`, k=64 (the whole row), fp32:**

```
device  v[0:4] = 7fc00000 7f800000 42700000 426c0000    i = 60 62 59 58
device  tail   = 3f800000 ff800000 ffc00000             i = 0 63 61
torch   v[0:4] = ffc00000 7fc00000 7f800000 42700000    i = 61 60 62 59
```

The device order is `+NaN > +inf > finite > -inf > -NaN` - a strict total order in which **`-NaN` is the unique minimum**. torch keeps both NaNs at the top (in either order) and `-inf` last.

**C. `largest=False`, k=16, same row** - this is the arm where the wrong selection is most visible, and where the returned NaN bits can be inspected:

```
[fp32] device v[0:4] = ffc00000 ff800000 3f800000 40000000   i = 61 63 0 1
[fp32] torch  v[0:4] = ff800000 3f800000 40000000 40400000   i = 63 0 1 2
[bf16] device v[0:4] = ff80     ff80     3f80     4000       i = 61 63 0 1
[bf16] torch  v[0:4] = ff80     3f80     4000     4040       i = 63 0 1 2
```

The fp32 device returns the `-NaN` lane **bit-exact as `0xffc00000`** at rank 0: the value is not substituted, the *ordering* is wrong. This is what rules out "the NaN is turned into an infinity of its own sign" and rules out a packer cause. The bf16 arm returns the same permutation with `ff80`, because the bf16 pack flushes NaN to a sign-following infinity. (For reference, `largest=False, k=64` on the device ends `42700000 7f800000 7fc00000` - the exact mirror of B, i.e. one single total order.)

**D. Second run, separate script, one NaN lane inside a `1..64` row, `k=16`, `largest=True`:**

| dtype | input NaN | device `i[0:4]` | torch `i[0:4]` | mismatches |
|---|---|---|---|---|
| bf16 | `7fc0` @3 | 3, 63, 62, 61 | 3, 63, 62, 61 | values 1/16, indices 0/16 |
| bf16 | `ffc0` @3 | 63, 62, 61, 60 | 3, 63, 62, 61 | **values 16/16, indices 16/16** |
| fp32 | `7fc00000` @3 | 3, 63, 62, 61 | 3, 63, 62, 61 | values 0/16, indices 0/16 |
| fp32 | `ffc00000` @3 | 63, 62, 61, 60 | 3, 63, 62, 61 | **values 16/16, indices 16/16** |

Lane position (3 vs 7) has no effect - the discriminator is the NaN's **sign bit**. The same holds for `k=4/8/16/32` and for `largest=False`.

### The ordering claim is a tested hypothesis, not a guess

Hypothesis: the network orders lanes by `(sign, magnitude)` of the raw value pattern with no NaN case, which predicts (i) `+NaN` strictly above `+inf`, (ii) `-NaN` strictly **below** `-inf` (not merely "equal to `-inf`"), and (iii) for `largest=False` the `-NaN` lane comes back bit-exact. All three are visible in B and C. The competing explanation "a NaN is ranked as an infinity carrying its own sign" predicts `-NaN` on a level with `-inf`, and (under substitution) `ff800000` in the fp32 output of C; both are contradicted.

## What this is not

* **Not** the bf16 packer family (#51557, #54381, #56414). The ordering anomaly is identical in fp32, where the same lanes come back bit-exact, so no bf16->bf16 identity pack is involved; and in bf16 the *indices* match the fp32 indices exactly while only the values are flushed. These are two independent layers, and this report is about the ordering layer only.
* **Not** an infinity-handling bug: `+inf`/`-inf` lanes are ranked correctly in both dtypes and both directions, and they come back bit-exact.
* **Not** already reported. I searched open and closed issues and PRs (titles and bodies) for topk/NaN/sign-magnitude combinations: `#56500`, `#55881`, `#55882` are topk test-infra / LLK transpose-guard tickets, `#56467` is the sampling stable-network stopgap, `#51470` is `maximum/minimum/clamp` (a single `SFPSWAP` is a permutation), `#52930` is the general LLK SFPU edge-case tracker (`0/0`, `RsqrtCompat`, `Erfinv`) - none of them is this op surface or this symptom. Current `test_topk.py` special-value coverage is `+inf`/`-inf`/bfp8 only.

## Static context (source, both trees)

* Released tree: `tt_metal/tt-llk/tt_llk_blackhole/common/inc/sfpu/ckernel_sfpu_topk.h` (1053 lines, bitonic, 63x `SFPSWAP`) contains **zero** occurrences of "nan" (case-insensitive), and so do all 16 topk kernel files under `ttnn/cpp/ttnn/operations/reduction/topk/`.
* Current main: the same header is now the fused-key implementation (1676 lines, `4daf16fc`) and still contains **zero** NaN handling - while it does carry a dedicated sweep for the other exceptional value, `_topk_canonicalize_negzero_value_tiles_` (the `-0.0` canonicalization). `topk.cpp` (`942b6f0d`), `topk_common_funcs.hpp` (`91f83666`) and `topk_final.cpp` (`0b4b6f48`) on main also have zero NaN hits.
* That asymmetry (an explicit `-0.0` canonicalize next to no NaN case) is why I suspect main is affected as well. It is inference from source, not a measurement - I could not run main here.

## Why it matters

MoE routing, sampling and attention all reduce with `ttnn.topk`. A NaN score is normally the signal that something upstream is broken; here it is silently routed to one extreme of the row when the sign bit is 0 and to the **opposite** extreme when it is 1, and in the latter case the op returns a different element set with shifted indices. Producing the negative sign is easy in fp32: `ttnn.neg` preserves the NaN payload and flips the sign bit (`+NaN -> -NaN`, bit-exact), so `topk(x)` and `topk(-x)` do not merely differ in encoding, they **select different elements**.

## Reproducer

```python
# ttsim: TT_METAL_SIMULATOR=$HOME/sim/libttsim_bh.so TT_METAL_SLOW_DISPATCH_MODE=1 \
#        TT_METAL_DISABLE_SFPLOADMACRO=1 python repro_topk_nan_order.py
# ASCII-only output. All NaN/Inf inputs are written as raw bits and read back as raw bits.
import torch
import ttnn

dev = ttnn.open_device(device_id=0)


def i32(b):
    return b - 4294967296 if b >= 2147483648 else b


def i16(b):
    return b - 65536 if b >= 32768 else b


def h32(t):
    return ["%08x" % (int(v) & 0xFFFFFFFF) for v in t.view(torch.int32).reshape(-1)]


def h16(t):
    return ["%04x" % (int(v) & 0xFFFF) for v in t.view(torch.int16).reshape(-1)]


def il(t):
    return [int(v) for v in t.reshape(-1)]


SPECIALS = [(60, 0x7FC00000, 0x7FC0), (61, 0xFFC00000, 0xFFC0),
            (62, 0x7F800000, 0x7F80), (63, 0xFF800000, 0xFF80)]

for is_bf16 in (True, False):
    tag = "bf16" if is_bf16 else "fp32"
    dt = torch.bfloat16 if is_bf16 else torch.float32
    hx = h16 if is_bf16 else h32
    x = torch.zeros(1, 1, 1, 64, dtype=dt)
    x[0, 0, 0, :60] = torch.arange(1, 61, dtype=dt)
    view = x.view(torch.int16) if is_bf16 else x.view(torch.int32)
    for lane, b32, b16 in SPECIALS:
        view[0, 0, 0, lane] = i16(b16) if is_bf16 else i32(b32)

    d = ttnn.from_torch(x.clone(), dtype=ttnn.bfloat16 if is_bf16 else ttnn.float32,
                        layout=ttnn.TILE_LAYOUT, device=dev)
    print("[%s] roundtrip lanes 60:64 = %s" % (tag, hx(ttnn.to_torch(d))[60:64]))

    for largest in (True, False):
        for k in ((64, 16) if largest else (16,)):
            tv, ti = ttnn.topk(d, k, dim=3, largest=largest)
            print("[%s] device largest=%-5s k=%-2d v[0:4]=%s i[0:4]=%s"
                  % (tag, largest, k, hx(ttnn.to_torch(tv))[:4], il(ttnn.to_torch(ti))[:4]))
            rv, ri = torch.topk(x, k, dim=3, largest=largest)
            print("[%s] torch  largest=%-5s k=%-2d v[0:4]=%s i[0:4]=%s"
                  % (tag, largest, k, hx(rv)[:4], il(ri)[:4]))
    if not is_bf16:
        for largest in (True, False):
            tv, ti = ttnn.topk(d, 64, dim=3, largest=largest)
            print("[fp32] device largest=%-5s k=64 tail v[-3:]=%s i[-3:]=%s"
                  % (largest, hx(ttnn.to_torch(tv))[-3:], il(ttnn.to_torch(ti))[-3:]))

ttnn.close_device(dev)
print("REPRO DONE (ttsim, no silicon)")
```

Output (verbatim, trimmed to the evidence lines):

```
[bf16] roundtrip lanes 60:64 = ['7fc0', 'ffc0', '7f80', 'ff80']
[bf16] device largest=True  k=64 v[0:4]=['7f80', '7f80', '4270', '426c'] i[0:4]=[60, 62, 59, 58]
[bf16] torch  largest=True  k=64 v[0:4]=['7fc0', '7fc0', '7f80', '4270'] i[0:4]=[61, 60, 62, 59]
[bf16] device largest=False k=16 v[0:4]=['ff80', 'ff80', '3f80', '4000'] i[0:4]=[61, 63, 0, 1]
[bf16] torch  largest=False k=16 v[0:4]=['ff80', '3f80', '4000', '4040'] i[0:4]=[63, 0, 1, 2]
[fp32] roundtrip lanes 60:64 = ['7fc00000', 'ffc00000', '7f800000', 'ff800000']
[fp32] device largest=True  k=64 v[0:4]=['7fc00000', '7f800000', '42700000', '426c0000'] i[0:4]=[60, 62, 59, 58]
[fp32] torch  largest=True  k=64 v[0:4]=['ffc00000', '7fc00000', '7f800000', '42700000'] i[0:4]=[61, 60, 62, 59]
[fp32] device largest=False k=16 v[0:4]=['ffc00000', 'ff800000', '3f800000', '40000000'] i[0:4]=[61, 63, 0, 1]
[fp32] torch  largest=False k=16 v[0:4]=['ff800000', '3f800000', '40000000', '40400000'] i[0:4]=[63, 0, 1, 2]
[fp32] device largest=True  k=64 tail v[-3:]=['3f800000', 'ff800000', 'ffc00000'] i[-3:]=[0, 63, 61]
[fp32] device largest=False k=64 tail v[-3:]=['42700000', '7f800000', '7fc00000'] i[-3:]=[59, 62, 60]
```

## Request

1. Please run the reproducer on silicon (Blackhole/Wormhole) and confirm the `-NaN` ordering. If a `-NaN` lane is **not** ranked below `-inf` on hardware, then this is a ttsim-model finding and I will take it to the simulator tracker instead.
2. Confirm the intended contract for NaN in `topk`. The docstring and the golden both say `torch.topk`, and #55346/#55350 settled the same question for `argmax`; if torch parity is intended, `topk` needs a NaN case (the existing `_topk_canonicalize_negzero_value_tiles_` sweep in the LLK header is the obvious place to put it alongside).
3. Whether current `main` (fused-key local-sort) is affected. My reading is that it is, from the source (no NaN handling anywhere in either tree), but I could not build/run main here, so please treat it as an inference.

Happy to produce full logs, the second script behind table D, or a patch arm that pins the ordering to a single instruction if that helps triage. I can also add the regression test if you want one (the current `test_topk.py` special-value coverage is `+inf`/`-inf`/bfp8 only).

**Comment by heroxcreed** [comment]
Hi @jberkowitzTT — I reported this and would like to take it. The bit-exact reproduction (device read-back, strict same-lane / same-k / same-largest controls) and the resulting ordering rule are in the issue body.
Could you assign the issue to me?

**Comment by heroxcreed** [comment]
Hi @jberkowitzTT — following up on my earlier request to take this one.

The defect is confirmed and I have a verified fix ready. I would like to open a PR against `tt-metal` — please tell me if that works for you.

**Root cause.** `TTI_SFPSWAP` compares words by sign and magnitude, and the bitonic sort network has no NaN branch. A `-NaN` lane (0xFFC00000) therefore sorts as a very large *negative* number: it is excluded from a `largest=True` top-k and is ranked as the minimum for `largest=False`. `+NaN` happens to survive only because its raw pattern already sits above `+inf`.

**Fix.** Canonicalize NaN in the freshly loaded value tiles before the local sort, in the same pass as the existing `-0.0` canonicalization: every NaN becomes `+NaN`, the one encoding that sign-magnitude order already places above `+inf`. No change to the compare-exchange network itself. The NaN test costs no extra instructions — for a 32-bit DEST, `((|x| >> 23) + 1) << 23` sets the top bit iff the exponent is all-ones, and `|x| << 9` is non-zero iff the mantissa is non-zero, so `±inf` and finite values stay out.

**Measured on ttsim (simulator; no silicon available to me).** Same input, same controls, fp32:
- `largest=True, k=16`: device `i[0:4] = [60,62,59,58]` before — i.e. the `-NaN` lane (61) is excluded from the top 4 while `+inf` (62) is kept — vs `[60,61,62,59]` after, with both NaN lanes ahead of `+inf` (torch's set for this input is `[61,60,62,59]`, so the NaN set now matches).
- `largest=False, k=16`: device `[61,63,0,1]` before — the `-NaN` lane ranked as the minimum, below `-inf` — vs `[63,0,1,2]` after, which is **bit-identical to `torch.topk`**.
- Regression companion (24 checks: no-NaN / 4-NaN / 16-NaN inputs, both directions, k=16 and k=64, ±inf sentinels): **8 failing before, 0 after**.
- Scope: `tt_metal/tt-llk/**` only — `tt_llk_blackhole` and `tt_llk_wormhole_b0` (`ckernel_sfpu_topk.h`). No op-level API change. Patch is `git apply --check` clean.

**One thing I did *not* fix, and I think you should know why.** On the bf16 path a NaN never reaches the sort at all: the device-side layout conversion is not value-preserving for bf16. On ttsim, `to_layout(ROW_MAJOR→TILE)` (and untilize, and bf16 typecast) rewrites every NaN payload as a same-signed `inf` — `7f81/7f82/7fa0/7fc0/7fe0/7fff` → `7f80` — while fp32 is bit-exact through the same paths and host-side packing is exact for both ([`tilize.cpp:32-35`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/kernel/compute/tilize.cpp) selects `Fp32Mode::Lossless` only for fp32 inputs, and `has_supported_fast_tilize_format()` includes `Float16_b`, so bf16 always takes the fast path). (The same experiment also flushes bf16 subnormals, but that part is expected — `tt_llk_*/common/inc/ckernel_defs.h` documents subnormals as unsupported by the hardware and rounded to zero, so I am not counting it against anything.) I cannot rule out that the NaN-payload rewrite is my simulator rather than silicon, so I have not written it up as a defect. I am happy to send a 40-line repro with the exact bit patterns if you want to look.

Thanks — and if the issue can be assigned to me, I'll have the PR up immediately.
